dill/_utils.py

```diff
@@ -0,0 +1,207 @@
+#!/usr/bin/env python
+#
+# Author: Leonardo Gama (@leogama)
+# Copyright (c) 2022 The Uncertainty Quantification Foundation.
+# License: 3-clause BSD.  The full license text is available at:
+#  - https://github.com/uqfoundation/dill/blob/master/LICENSE
+"""auxiliary internal classes used in multiple submodules, set here to avoid import recursion"""
+
+__all__ = ['AttrDict', 'ExcludeRules', 'Filter', 'RuleType']
+
+import logging
+logger = logging.getLogger('dill._utils')
+
+class AttrDict(dict):
```

Comment:
I don't love the use of this.  To me, it seems like additional overhead.

Reply:
This one was already there, just moved it to `_utils.py`. Currently, it's only used for the "module map" in `_stash_modules()` and related functions, and only `__getattr__()` is actually called. I added the `set` and `del` methods for completeness, with safeguards against, for example, overwriting the `items()` method with a dict key. For the current use, I would guess the overhead negligible, as attribute lookup is a heavily optimized operation.